### DISCUSSION TITLE
Win32 Xaml Island load component from library

### DISCUSSION THREAD

**Discussion by nbkalex** [microsoft/WindowsAppSDK-Samples#312] [General] [answered]
There is a sample for WinUI 2 for doing it [here](https://github.com/microsoft/Xaml-Islands-Samples/tree/master/Samples/Win32/SampleCppApp) and a how to [here](https://learn.microsoft.com/en-us/windows/apps/desktop/modernize/host-custom-control-with-xaml-islands-cpp).

Is there a way to do it with WindowsAppSDK?

**Comment by DarranRowe** [comment, marked as answer]
It mostly works the same way.
The biggest differences are that you need to create a dispatcher queue (Microsoft.UI.Dispatching.DispatcherQueue via Microsoft.UI.Dispatching.DispatcherQueueController) on the thread which has the message pump first.
The islands use the window id (Microsoft.UI.WindowId) instead of the interop interfaces. This means that to initialise the DesktopWindowXamlSource instance, you would first use the GetWindowIdFromWindow function (C++) or Win32Interop.GetWindowIdFromWindow static method (C#) and then call DesktopWindowXamlSource.Initialize using this WindowId, instead of obtaining the IDesktopWindowXamlSourceNative interface. This interface no longer exists in the WinUI 3 version of Xaml Islands.
To work with the content window itself, you would get the underlying site using DesktopWindowXamlSource.SiteBridge and then use Microsoft.UI.Content.DesktopSiteBridge.WindowId to get the window id for that underlying content window. You would then use the GetWindowFromWindowId function (C++) or Win32Interop.GetWindowFromWindowId static method (C#) to get the underlying HWND. This is used instead of IDesktopWindowXamlSourceNative interface.
The PreTranslateMessage member of the IDesktopWindowXamlSourceNative2 interface has been replaced by a single call to the ContentPreTranslateMessage function. It works in the same way, if you call the function and it returns true, then you shouldn't then call DispatchMessage.
Besides this, you should be able to take what works with WinUI 2 and just do the namespace changes needed to convert to WinUI 3. I haven't tested if working completely without the Application instance works if you don't want to load the control styles though.

**Reply by nbkalex** [reply]
Thanks Darran.

After doing these steps and many hours of configuring, compiling, linking and debugging everything is working fine :)
There were some internal misleading crashes that were making the debug experience terrible.

Another issue I'm facing is that I'm getting a lot of noise using **_CRTDBG_LEAK_CHECK_DF** for detecting memory leaks.
As @jevansaks says in [this ](https://github.com/microsoft/microsoft-ui-xaml/issues/3973) issue, these are not a signs of runtime memory leaks. My problem is that it's getting harder catching the memory leaks from my source code. Do you have suggestions on this?
